### PR TITLE
Support selecting from boot from volume and image

### PR DESCRIPTION
**Pull request by hkumarmk**
This patch allow one to select booting a node from volume and image by changing
"boot_volume" to a name which is mapped in the map file under volumes section.

Also changed nova boot from using block_device_mapping after cinder volume creation
to block_device_mapping_v2, which will do both creating volume and boot from the
volume. So overcast do not need to manage the volume separately, but nova will
handle creating/deleting volume along with the node.

### CODE REVIEW

**Comment by sorenh** [comment]
Is there any benefit to doing it this way (requiring the user to specify this in the mappings file) as opposed to making the decision based on whether there's a block storage endpoint (and/or whether the os-boot-from-volume extension is available in Nova)?

File: overcast/runner/__init__.py [hunk 1]
```diff
@@ -152,6 +152,19 @@ def __init__(self, name, info, runner, keypair=None, userdata=None):
...
+        if self.info.get('boot_volume') in self.runner.mappings.get('volumes',{}):
+            self.info['boot_volume'] = self.runner.mappings['volumes'][self.info['boot_volume']]
+
+        ##
+        # This is to allow disallowing boot_volume, in case mapfile dont have
+        # any such settings and environment file have a reference name for
+        # boot_volume
+        ##
+        try:
+            just_tmp_value=int(self.info['boot_volume'])
```
Comment (sorenh) [inline overcast/runner/__init__.py]: If you're simply trying to validate that `self.info['boot_volume']` is an integer (in a string), don't bother assigning it to a variable.

File: overcast/runner/__init__.py [hunk 2]
```diff
@@ -152,6 +152,19 @@ def __init__(self, name, info, runner, keypair=None, userdata=None):
...
+        except:
```
Comment (sorenh) [inline overcast/runner/__init__.py]: Please use an explicit list of exceptions that you want to catch. In this case, it's probably only ValueError?

File: overcast/runner/__init__.py [hunk 3]
```diff
@@ -202,18 +215,17 @@ def build(self):
...
-
-        bdm = {'vda': '%s:::1' % (volume.id,)}
+        if self.info['boot_volume']:
+            bdm_v2 = [{'source_type': 'image', 'uuid': self.info['image'],
+                        'destination_type': 'volume', 'volume_size': self.info['boot_volume'],
+                        'delete_on_termination': 'true', 'boot_index': '0'}]
+            image = None
+        else:
+            image = self.info['image']
+            bdm_v2 = None
```
Comment (sorenh) [inline overcast/runner/__init__.py]: bdm_v2 does not work on HP Cloud, so please change this back.

Comment (sorenh) [inline overcast/runner/__init__.py]: Hm, let me rephrase: *I* could not get bdmv2 to work on HP Cloud. If HP Cloud now supports bdmv2 and/or you've found a way to make it work, that's fine, but it needs to work there, too.

File: overcast/tests/runner/test_main.py [hunk 4]
```diff
@@ -682,6 +729,59 @@ def status(self):
         self.dr.record_resource.assert_any_call('port', 'nicuuid2')
         self.dr.record_resource.assert_any_call('server', 'serveruuid')
 
+
+    @mock.patch('overcast.runner.DeploymentRunner.create_port')
+    @mock.patch('overcast.runner.DeploymentRunner.get_nova_client')
+    @mock.patch('overcast.runner.DeploymentRunner.get_neutron_client')
+    @mock.patch('overcast.runner.DeploymentRunner.get_cinder_client')
+    @mock.patch('overcast.runner.time')
+    def test_create_node_with_map_boot_from_image(self, time, get_cinder_client, get_neutron_client, get_nova_client, create_port):
```
Comment (sorenh) [inline overcast/tests/runner/test_main.py]: There's far too much duplicated code in these three tests.

Try to put all the common code into a new method. Only methods that start with `test_` are run during the unit test run, so you can just call it e.g. `_test_create_node`. Make it take some extra arguments that makes it test the different scenarios. At a glance, it seems like you could simply pass in three variables: a dict that would be used to extend the node info, another dict representing the expected bdm dict being passed in and a string/NoneType denoting the expected image being passed to nc.servers.create.

Does that make sense?

File: overcast/runner/__init__.py (hunk 1 above)
Comment (hkumarmk) [inline overcast/runner/__init__.py]: when I tried it from python console , I see that was printing it, I will change it

File: overcast/runner/__init__.py (hunk 2 above)
Comment (hkumarmk) [inline overcast/runner/__init__.py]: oh yeah ill do it :)

File: overcast/runner/__init__.py (hunk 3 above)
Comment (hkumarmk) [inline overcast/runner/__init__.py]: I dont know that, I may have to test it (unfortunately I dont have an account to test it)

File: overcast/tests/runner/test_main.py (hunk 4 above)
Comment (hkumarmk) [inline overcast/tests/runner/test_main.py]: yes pretty much :) ill do it... sorry I am very new in this.

**Comment by hkumarmk** [comment]
We can check the block endpoint available or not to decide whether to use boot from volume or not (so if one specified a volume size, still it can use the endpoint availablility to decide whether to boot from volume or not), but still i feel it make sense to make it configurable in case people still want to boot from disk if both options are available.

And may be I would just fail in case somebody choose to boot from volume against a cloud which dont have a volume endpoint rather than switching automatically to disk - especially If I think this as a generic provisioning code.

**Comment by sorenh** [comment]
Hm. Ok.

At any rate, I'm not too happy about the config changes. I would expect a `[volumes]` section to deal with actual volumes, not just the size of volumes to be created.

I'm a bit torn on the subject, really. I would love to be able to list volumes in the stack descriptions. These volumes could hold important data that should not be tied to a single instance. Such a section in the stack would most likely be called `volumes` and the mappings for them would go into a `[volumes]` section in the mappings file.

Do you have any ideas for how we can accomodate that use case as well as what you're trying to do here?

**Comment by hkumarmk** [comment]
I will check this

**Comment by hkumarmk** [comment]
Fixed other problems (fixing test code, and other small issues), bdm_v2 and the configuration issue will be fixed next.

File: overcast/runner/__init__.py (hunk 3 above)
Comment (hkumarmk) [inline overcast/runner/__init__.py]: I will try to fix it along with the volume config fix

File: overcast/tests/runner/test_main.py (hunk 4 above)
Comment (hkumarmk) [inline overcast/tests/runner/test_main.py]: This is done

**Comment by sorenh** [comment]
If/when we add an OpenStack driver again, we can look into this. Closing it for now.
